### PR TITLE
refactor executor::cancel to use a spinning state

### PR DESCRIPTION
**Pull request by wjwwood**
@jacquelinekay as we discussed off-line.

### CODE REVIEW

File: rclcpp/include/rclcpp/executor.hpp [hunk 1]
```diff
@@ -180,9 +180,8 @@ class Executor
     return FutureReturnCode::INTERRUPTED;
   }
 
-  /// Stop everything
```
Comment (jacquelinekay) [inline rclcpp/include/rclcpp/executor.hpp]: Haha, I see you didn't find this comment every descriptive

Comment (wjwwood) [inline rclcpp/include/rclcpp/executor.hpp]: I've just been trying to be better about code docs in general. :smile:

**Comment by jacquelinekay** [comment]
Looks good at first glance. Can you add a test case, maybe in the other PR to system_tests, with two threads that try to enter `spin` at the same time?

**Comment by wjwwood** [comment]
Yup, but should I do it on the `cancel` branch or my `cancel_wjwwood` branch?

**Comment by jacquelinekay** [comment]
Either is fine.

**Comment by wjwwood** [comment]
I push a commit to this pr with the new tests: https://github.com/ros2/system_tests/pull/73 (https://github.com/ros2/system_tests/commit/3b346be02cb127d7a88daa081427905637531725)

I'll leave it to you to merge this branch.

File: rclcpp/include/rclcpp/executor.hpp [hunk 2]
```diff
@@ -180,9 +180,8 @@ class Executor
...
-  /**
-   */
+  /// Cancels any running spin* function, causing it to return.
```
Comment (dirk-thomas) [inline rclcpp/include/rclcpp/executor.hpp]: `Cancel` instead of `Cancels`.

**Pull request merged**